students/TimPauley/session05/MailRoomPart3.py

```diff
@@ -0,0 +1,153 @@
+#code to handle errors better, such as when a user inputs bad data.
+#3. Comprehensions- Can you use comprehensions to clean up your code a bit?
+#4. Note: you may be tempted to replace loops like this:
+
+#Part 2
+#This is the update from Part 2
+#Update using Dictionaries
+#1. See if you can use a dict to switch between the user’s selections.
+#2. See if you can use a dict to switch between the users selections. 
+#3. see Using a Dictionary to switch for what this means.
+#4. Convert your main donor data structure to be a dict.
+#5. Try to use a dict and the .format() method to produce the letter as one 
+#big template, rather than building up a big string that produces the letter in parts.
+
+import sys
+import math 
+
+# this creates better printing
+from textwrap import dedent
+
+#Get Donors
+def get_donor_db():
+    return {'william gates iii': ("William Gates III", [653772.32, 12.17]),
+            'jeff bezos': ("Jeff Bezos", [877.33]),
+            'paul allen': ("Paul Allen", [663.23, 43.87, 1.32]),
+            'mark zuckerberg': ("Mark Zuckerberg", [1663.23, 4300.87, 10432.0]),
+            }
+
+#List out donors
+def list_donors():
+    listing = ["Donor list:"]
+    for donor in donor_db.values():
+        listing.append(donor[0])
+    return "\n".join
+
+# this fuction finds a donors name
+def find_donor_Name():
+    name = name.strip()
+    donor = (name, [])
+    donor_db[name.lower()] = donor
+    return donor
+    
+def main_menu_selection():
+    action = input(dedent('''
+        Choose an action:
+        1 - Send a Thank You
+        2 - Create a Report
+        3 - Send letters to everyone
+        4 - Quit
+        > '''))
+    return action.strip()
+
+def gen_letter(donor):
+    return dedent('''Hey {0:s},
+          Thanks for yur donation of ${1:.2f}.
+          It will help with our mission.
+                         Thanks,
+                            -The Dream Team
+          '''.format(donor[0], donor[1][-1]))
+
+#1 - Send Thank-you card
+def send_thank_you():
+    while True:
+        name = input("Enter a donor's name (Enter 'list' to see all donors or 'menu' to exit)> ").strip()
+        if name == "list":
+            print(list_donors())
+        elif name == "menu":
+            return
+        else:
+            break    
+        while True:
+            amount_str = input("Enter a donation amount (or 'menu' to exit)> ").strip()
+        if amount_str == "menu":
+            return
+        try:
+            amount = float(amount_str)
+        #Eception Handling: check to see if value is numeric
+            if math.isnan(amount) or math.isinf(amount) or round(amount, 2) == 0.00:
+                raise ValueError
+        except ValueError:
+                print("error: donation amount is invalid\n")
+        else:
+            break
+            donor = find_donor(name)
+        if donor is None:
+            donor = add_donor(name)
+    donor[1].append(amount)
```

Comment:
Getting an error here "UnboundLocalError: local variable 'donor' referenced before assignment" . Reach out if you have trouble resolving.